### PR TITLE
[TF2XLA] Add EuclideanNorm kernel

### LINKED ISSUE #41915: tf.math.reduce_euclidean_norm can not be in tf.function with experimental_compile=True
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): colab
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.4.0-dev20200730
- Python version: colab
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory: no

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

```
EuclideanNorm: unsupported op: No registered 'EuclideanNorm' OpKernel for XLA_CPU_JIT devices compatible with node {{node EuclideanNorm}}
```

**Describe the expected behavior**

The op can run as usual.

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```python3
import tensorflow as tf

x = tf.complex(tf.random.uniform(shape=(5, 5)), tf.random.uniform(shape=(5, 5)))

@tf.function(experimental_compile=True)
def reduce_euclidean_norm(x):
  return tf.math.reduce_euclidean_norm(x)

print(reduce_euclidean_norm(x))
```

https://colab.research.google.com/drive/1YENmpGDLU6kCupseizWTO1wBiDDj8v9x?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

### PR DESCRIPTION
**Pull request by WindQAQ**
Fixes #41915.

### CODE REVIEW

File: tensorflow/compiler/tf2xla/kernels/reduction_ops.cc [hunk 1]
```diff
@@ -184,5 +187,44 @@ class AnyOp : public XlaReductionOp {
...
+  void BuildReducer(xla::XlaBuilder* builder, const xla::XlaOp& scalar_lhs,
+                    const xla::XlaOp& scalar_rhs) override {
+    xla::Add(scalar_lhs, scalar_rhs);
+  }
+
+  xla::XlaOp BuildFinalizer(
+      xla::XlaBuilder* /*builder*/, const xla::XlaOp& input,
+      const xla::XlaOp& reduce_output,
+      const std::vector<int64>& dimensions_to_reduce) override {
+    if (xla::primitive_util::IsIntegralType(xla_reduction_type_)) {
```
Comment (cheshire) [inline tensorflow/compiler/tf2xla/kernels/reduction_ops.cc:213]: Is it a good idea to do that? Reduction in floating point numbers might yield different numbers.
@sanjoy WDYT?

File: tensorflow/compiler/tf2xla/kernels/reduction_ops.h [hunk 2]
```diff
@@ -39,6 +39,10 @@ class XlaReductionOp : public XlaOpKernel {
   // Return the base case for the reduction.
   virtual xla::XlaOp InitialValue(xla::XlaBuilder* builder) = 0;
 
+  // Processes data before reduction.
+  virtual xla::XlaOp ProcessData(xla::XlaBuilder* builder,
+                                 const xla::XlaOp& data);
+
```
Comment (sanjoy) [inline tensorflow/compiler/tf2xla/kernels/reduction_ops.h:45]: Should be called `PreprocessInput` or something like that.  `ProcessData` sounds super generic.

File: tensorflow/compiler/tf2xla/kernels/reduction_ops.cc (hunk 1 above)
Comment (sanjoy) [reply tensorflow/compiler/tf2xla/kernels/reduction_ops.cc:213]: I don't think this is reducing in floating point, it is just doing the final sqrt in floating point.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2xla/kernels/reduction_ops.cc (hunk 1 above)
Comment (WindQAQ) [reply tensorflow/compiler/tf2xla/kernels/reduction_ops.cc:213]: @sanjoy is right, it's the finalizer that processes output of reduction.

**Review by cheshire — DISMISSED** [review]

File: tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc [hunk 3]
```diff
@@ -35,6 +35,12 @@ XlaReductionOp::XlaReductionOp(OpKernelConstruction* ctx,
       ctx, DataTypeToPrimitiveType(reduction_type_, &xla_reduction_type_));
 }
 
+// The default pre-processor directly returns the data. This can be overridden.
+xla::XlaOp XlaReductionOp::ProcessInput(xla::XlaBuilder* /*builder*/,
```
Comment (sanjoy) [inline tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc:39]: The verb here needs to be more specific than `Process`.  `Preprocess` is fine.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc (hunk 3 above)
Comment (WindQAQ) [reply tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc]: Updated.

File: tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc [hunk 4]
```diff
@@ -122,7 +129,7 @@ void XlaReductionOp::Compile(XlaOpKernelContext* ctx) {
   xla::XlaComputation reduction_computation = r.Build().ConsumeValueOrDie();
 
   auto reduce = xla::Reduce(data, initial, reduction_computation, xla_axes);
-  auto finalized = BuildFinalizer(b, data, reduce, xla_axes);
+  auto finalized = BuildFinalizer(b, ctx->Input(0), reduce, xla_axes);
```
Comment (sanjoy) [inline tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc:132]: Why did this need to change?  Earlier we were passing in the result of `ConvertElementType` right?

**Review by sanjoy — CHANGES_REQUESTED** [review]
One minor comment, LGTM otherwise.

File: tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc (hunk 4 above)
Comment (WindQAQ) [reply tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc:132]: You're right. My original thought is to make codes unchanged in this part, so maybe we should do something like
```cpp
  auto input = xla::ConvertElementType(ctx->Input(0), type);
  auto data = PreprocessInput(b, input);
  ...
  auto finalized = BuildFinalizer(b, input, reduce, xla_axes);
```
What do you think about this? Thanks!

Comment (sanjoy) [reply tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc:132]: LGTM, except that I'd prefer s/input/converted_input/

File: tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc [hunk 5]
```diff
@@ -111,7 +117,8 @@ void XlaReductionOp::Compile(XlaOpKernelContext* ctx) {
   xla::PrimitiveType type;
   TF_CHECK_OK(DataTypeToPrimitiveType(reduction_type_, &type));
 
-  auto data = xla::ConvertElementType(ctx->Input(0), type);
+  auto input = xla::ConvertElementType(ctx->Input(0), type);
+  auto converted_input = PreprocessInput(b, input);
```
Comment (sanjoy) [inline tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc:121]: Sorry for the miscommunication, I meant:

```
  auto converted_input = xla::ConvertElementType(ctx->Input(0), type);
  auto preprocessed_input = PreprocessInput(b, converted_input);
```

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc (hunk 5 above)
Comment (WindQAQ) [reply tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc]: No worries 😃 Thanks for the review BTW.

**Review by sanjoy — APPROVED** [review]

**Comment by cheshire** [comment]
@WindQAQ This does not seem to work for complex numbers (actually I'm not sure why), but e.g. `EuclideanNormReductionTest.testComplex64` is failing in `//tensorflow/python/kernel_tests:reduction_ops_test_xla_gpu`. Would you be willing to investigate?

**Comment by WindQAQ** [comment]
@Cheshire will do. Can you share the full failing log? Thank you!

**Comment by cheshire** [comment]
@WindQAQ It produces a completely different output:

```
Not equal to tolerance rtol=1e-06, atol=1e-06
Mismatched value: a is different from b.
not close where = (array([1]),)
not close lhs = [2.236068+0.j]
not close rhs = [1.-2.j]
```

I'm trying to figure out what is going on, surprisingly, the HLO looks already constant folded (even before optimizations).

**Comment by cheshire** [comment]
@WindQAQ I have found the bug: in case of empty axis,  the base class returns output: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc;l=52-58?q=XlaReductionOp::Compile

In your case, it might be that the finalizer still has to be applied.

To me it indicates that subclassing is not a correct approach here: euclidean norm is conceptually not a reduction, because it's not an identity for a singular element. Could you wrap it instead?

**Comment by WindQAQ** [comment]
> @WindQAQ I have found the bug: in case of empty axis, the base class returns output: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/tf2xla/kernels/reduction_ops_common.cc;l=52-58?q=XlaReductionOp::Compile
> 
> In your case, it might be that the finalizer still has to be applied.
> 
> To me it indicates that subclassing is not a correct approach here: euclidean norm is conceptually not a reduction, because it's not an identity for a singular element. Could you wrap it instead?

Thanks for pointing out this! I'll prepare a PR for it.

**Pull request merged**